src/rgw/rgw_keystone.cc

```diff
@@ -241,6 +241,75 @@ int KeystoneService::get_keystone_admin_token(CephContext * const cct,
   return 0;
 }
 
+int KeystoneService::get_keystone_barbican_token(CephContext * const cct,
+                                              std::string& token)
+{
+  std::string token_url;
+
+  if (get_keystone_url(cct, token_url) < 0) {
+    return -EINVAL;
+  }
+
+  KeystoneToken t;
+
+  /* Try cache first. */
+  if (RGWKeystoneTokenCache::get_instance().find_barbican(t)) {
+    ldout(cct, 20) << "found cached barbican token" << dendl;
+    token = t.token.id;
+    return 0;
+  }
+
+  bufferlist token_bl;
+  RGWKeystoneHTTPTransceiver token_req(cct, &token_bl);
+  token_req.append_header("Content-Type", "application/json");
+  JSONFormatter jf;
+
+  const auto keystone_version = KeystoneService::get_api_version();
+  if (keystone_version == KeystoneApiVersion::VER_2) {
+    KeystoneBarbicanTokenRequestVer2 req_serializer(cct);
+    req_serializer.dump(&jf);
+
+    std::stringstream ss;
+    jf.flush(ss);
+    token_req.set_post_data(ss.str());
+    token_req.set_send_length(ss.str().length());
```

Comment:
`std::stringstream::str()` returns a copy of its internal string by value, so avoid calling it twice here:
```c++
    auto str = ss.str();
    token_req.set_post_data(str);
    token_req.set_send_length(str.length());
```